### PR TITLE
[CLOUD-268] Repo-centric permission syncing with GitHub App

### PR DESCRIPTION
**Pull request by unknwon**
This PR allows repo-centric permission syncing to work with GitHub App. A new Installation Access Token is created for every sync.

## Test plan

Added and updated unit tests.

Manually tested in local environment to confirm permission sync is working.

### CODE REVIEW

File: enterprise/internal/authz/github/app.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	"net/url"
+	"time"
+
+	"github.com/sourcegraph/sourcegraph/internal/extsvc"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/auth"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/github"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// newAppProvider creates a new GitHub Provider with a client that creates a new installation access token.
```
Comment (unknwon) [inline enterprise/internal/authz/github/app.go:15]:
```suggestion
// newAppProvider creates a new authz Provider for GitHub App.
```

File: enterprise/internal/authz/github/app.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	apiURL, _ := github.APIRoot(baseURL)
+	appClient := github.NewV3Client(apiURL, auther, nil)
+	return &Provider{
+		urn:      urn,
+		codeHost: extsvc.NewCodeHost(baseURL, extsvc.TypeGitHub),
+		client: func() (client, error) {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+			defer cancel()
+
+			// TODO(cloud-saas): Cache the installation access token until it expires.
```
Comment (unknwon) [inline enterprise/internal/authz/github/app.go:36]:
```suggestion
			// // TODO(cloud-saas): Cache the installation access token until it expires, see
			// https://sourcegraph.atlassian.net/browse/CLOUD-255
```

**Review by unknwon — COMMENTED** [review]
Nothing noticeable off to me (did not do a complete review), let's mark it ready for review!

File: enterprise/internal/authz/github/app.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	apiURL, _ := github.APIRoot(baseURL)
+	appClient := github.NewV3Client(apiURL, auther, nil)
+	return &Provider{
+		urn:      urn,
+		codeHost: extsvc.NewCodeHost(baseURL, extsvc.TypeGitHub),
+		client: func() (client, error) {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+			defer cancel()
+
+			// // TODO(cloud-saas): Cache the installation access token until it expires, see
```
Comment (unknwon) [inline enterprise/internal/authz/github/app.go:36]:
```suggestion
			// TODO(cloud-saas): Cache the installation access token until it expires, see
```

File: enterprise/internal/authz/github/app_test.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+				w.Header().Set("Content-Type", "application/json")
+				w.Write(respJson)
+			} else {
+				srvHit = true
+				gotHeader := r.Header.Get("Authorization")
+				wantHeader := "Bearer 1234"
+				if gotHeader != wantHeader {
+					t.Fatalf("wanted authorization header %s, got %s", wantHeader, gotHeader)
+				}
```
Comment (unknwon) [inline enterprise/internal/authz/github/app_test.go:42]: We're trading towards using `assert` and `require` packages.

```suggestion
				assert.Equal(t, wantHeader, gotHeader)
```

File: enterprise/internal/authz/github/app_test.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+				tokenString := "1234"
+				tokenExpiry := time.Now()
+				token := github.InstallationToken{
+					Token:        &tokenString,
+					ExpiresAt:    &tokenExpiry,
+					Permissions:  &github.InstallationPermissions{},
+					Repositories: []*github.Repository{},
+				}
+
+				respJson, _ := json.Marshal(token)
```
Comment (unknwon) [inline enterprise/internal/authz/github/app_test.go:32]: nit:
```suggestion
				respJSON, err := json.Marshal(token)
				require.NoError(t, err)
```

File: enterprise/internal/authz/github/app_test.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+			GitHubConnection: &schema.GitHubConnection{
+				Url: srv.URL,
+				Authorization: &schema.GitHubAuthorization{
+					GroupsCacheTTL: 72,
+				},
+				GithubAppInstallationID: "1234",
+			},
+		}
+
+		baseURL, _ := url.Parse(ghConnection.Url)
```
Comment (unknwon) [inline enterprise/internal/authz/github/app_test.go:57]: nit:

```suggestion
		baseURL, err := url.Parse(ghConnection.Url)
		require.NoError(t, err)
```

File: enterprise/internal/authz/github/app_test.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+		const bogusKey = `LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQpNSUlCUEFJQkFBSkJBUEpIaWprdG1UMUlLYUd0YTVFZXAzQVo5Q2VPZUw4alBESUZUN3dRZ0tabXQzRUZxRGhCCk93bitRVUhKdUs5Zm92UkROSmVWTDJvWTVCT0l6NHJ3L0cwQ0F3RUFBUUpCQU1BK0o5Mks0d2NQVllsbWMrM28KcHU5NmlKTkNwMmp5Nm5hK1pEQlQzK0VvSUo1VFJGdnN3R2kvTHUzZThYUWwxTDNTM21ub0xPSlZNcTF0bUxOMgpIY0VDSVFEK3daeS83RlYxUEFtdmlXeWlYVklETzJnNWJOaUJlbmdKQ3hFa3Nia1VtUUloQVBOMlZaczN6UFFwCk1EVG9vTlJXcnl0RW1URERkamdiOFpzTldYL1JPRGIxQWlCZWNKblNVQ05TQllLMXJ5VTFmNURTbitoQU9ZaDkKWDFBMlVnTDE3bWhsS1FJaEFPK2JMNmRDWktpTGZORWxmVnRkTUtxQnFjNlBIK01heFU2VzlkVlFvR1dkQWlFQQptdGZ5cE9zYTFiS2hFTDg0blovaXZFYkJyaVJHalAya3lERHYzUlg0V0JrPQotLS0tLUVORCBSU0EgUFJJVkFURSBLRVktLS0tLQo=`
+
+		provider, _ := newAppProvider(ghConnection.URN, baseURL, "1234", bogusKey, 1234)
+
+		cli, _ := provider.client()
```
Comment (unknwon) [inline enterprise/internal/authz/github/app_test.go:63]: ditto.

File: enterprise/internal/authz/github/app_test.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+		// call any endpoint so that test server can check Authorization header
+		cli.ListTeamMembers(context.Background(), "anyOwner", "anyTeam", 0)
+
+		if !srvHit {
+			t.Fatal("did not hit server endpoint")
+		}
```
Comment (unknwon) [inline enterprise/internal/authz/github/app_test.go:70]:
```suggestion
		assert.True(t, srvHit, "hit server endpoint")
```

File: enterprise/internal/authz/github/authz_test.go [hunk 9]
```diff
@@ -153,5 +154,50 @@ func TestNewAuthzProviders(t *testing.T) {
...
+			)
+			if len(providers) != 1 || providers[0] == nil {
+				t.Fatal("expected a provider")
+			}
+			if len(problems) != 0 {
+				t.Fatalf("unexpected problems: %+v", problems)
+			}
+			if len(warnings) != 0 {
+				t.Fatalf("unexpected warnings: %+v", warnings)
+			}
```
Comment (unknwon) [inline enterprise/internal/authz/github/authz_test.go:200]: nit: use `assert` package

**Review by unknwon — COMMENTED** [review]
I can't approve my own PR, but ✅

File: enterprise/internal/authz/github/authz_test.go [hunk 10]
```diff
@@ -23,15 +25,10 @@ func TestNewAuthzProviders(t *testing.T) {
...
-		if len(problems) != 0 {
-			t.Fatalf("unexpected problems: %+v", problems)
-		}
-		if len(warnings) != 0 {
-			t.Fatalf("unexpected warnings: %+v", warnings)
-		}
+
+		assert.True(t, len(providers) == 0, "unexpected a providers: %+v", providers)
+		assert.True(t, len(problems) == 0, "unexpected problems: %+v", problems)
+		assert.True(t, len(warnings) == 0, "unexpected warnings: %+v", warnings)
```
Comment (unknwon) [inline enterprise/internal/authz/github/authz_test.go:31]: You can use `assert.Len` 😁

**Review by ryanslade — COMMENTED** [review]
Took a quick look and looks good to me, but I don't know enough about the context to actually approve. Will defer to @unknwon on that

**Comment by unknwon** [comment]
@pjlast You would need make an approval if you think it is good to merge, or wait one other team member from the Cloud SaaS.

**Review by pjlast — APPROVED** [review]

**Pull request merged**